### ISSUE TITLE
AML grabbing broken on Vero4k after PR #f2eef4e 

### ISSUE THREAD

**Issue by hissingshark** [hyperion-project/hyperion.ng#559]
Hi there,

**Issue:**
Hyperion.ng is no longer grabbing for video via AML-Grabber on the Vero4k.  The framebuffer grabs from the Kodi menu/OSD continue to work as normal.

---
**Observations:**
1. Debug output for Kodi menus is the expected:
> [hyperiond AMLOGICGRABBER] <INFO> FB mode

The live preview in the webGUI matches the TV output.

2. When starting video playback debug output is:
> [hyperiond AMLOGICGRABBER] <INFO> VPU mode
> [hyperiond AMLOGICGRABBER] (ERROR) Read of device failed: 11 - Resource temporarily unavailable

The live preview is black.  No LEDs.

Interestingly - if you skip forward/back in the video a single frame is displayed in the preview, corresponding with the point skipped from.

---
**Work done:**
I've bisected this back to [Amlogic: implement CAP_FLAG_AT_END mode ](https://github.com/hyperion-project/hyperion.ng/commit/f2eef4eceac7a1bd28aa767b6a128084e8e98e84)

I've tried with no effect all config combinations of:
Type (auto, amlogic, framebuffer)
Device (/dev/fb0, /dev/amvideocap0)

---
Many thanks in advance for any help with this.

**Comment by hissingshark** [comment]
I added some extra debugging lines to walk though what was happening.

Config changes didn't make any difference as the devices are hard coded:
https://github.com/hyperion-project/hyperion.ng/blob/b1b2079f972330c757b50f4c6615e810df4e3e10/libsrc/grabber/amlogic/AmlogicGrabber.cpp#L20-L21

The QThread timing method seems to work because the calls to:
`AmlogicGrabber::grabFrame`
are happening.
`AmlogicGrabber::isVideoPlaying`
successfully detects the start of video playback, then:
`AmlogicGrabber::grabFrame_amvideocap`
basically hangs at:
https://github.com/hyperion-project/hyperion.ng/blob/b1b2079f972330c757b50f4c6615e810df4e3e10/libsrc/grabber/amlogic/AmlogicGrabber.cpp#L144-L145
Here it either times out eventually with `pread` returning`-1`:
> [hyperiond AMLOGICGRABBER] (ERROR) Read of device failed: 11 - Resource temporarily unavailable



Or if during that time the video is skipped through,  then as described originally a frame will be successfully captured and displayed in the preview.

Simply disabling the new option
https://github.com/hyperion-project/hyperion.ng/blob/b1b2079f972330c757b50f4c6615e810df4e3e10/libsrc/grabber/amlogic/AmlogicGrabber.cpp#L134
Results in normal grabbing behavior, but will defeat the purpose of the commit (fixing the sync issue).

`AMVIDEOCAP` is enabled in the kernel here on OSMC, with a present `/dev/amvideocap0` and clearly capturing can work on the platform.

@Portisch 
What device / OS / kernel were you successful testing these changes on? I'm assuming there is some environmental difference here.
Any suggestions what may be different?  Is there anything further I can do to debug this issue?

Many thanks,

**Comment by Portisch** [comment]
This "slow down" Hyperion.NG fix for Amlogic does work only for CoreELEC systems!

**Comment by hissingshark** [comment]
Ah, thought so.  So there were changes at the "receiving end" too.  For CoreELEC I now see your handywork:
https://github.com/CoreELEC/linux-amlogic/commit/aa7632313f7e9161733409909073346017ecfde8
Many thanks for the pointer.

**Comment by AnthraxLeprosyPi** [comment]
Hi @hissingshark,

is there any fix for the vero yet? After updating to osmc 5-2 the sad face haunts me :-(

cheers
Anthrax

**Comment by hissingshark** [comment]
@AnthraxLeprosyPi 
Yep.  As it happens I posted [this morning on the OSMC forum](https://discourse.osmc.tv/t/osmc-and-hyperion/23293/537?u=hissingshark) once I had received my June update and confirmed it was working.

I had to implement the kernel changes that Portisch had made into OSMC (thanks again for the pointers @Portisch, the fix is working very well).  The June update brings in the underlying changes, so then you just need to update Hyperion.ng.  I couldn't update Hyperion.ng in my installer before hand because as per this issue thread it would have broken grabbing altogether without the kernel updates to go with it!

It's all in that post. I hope it works out for you!
